contracts/evm/src/eth/utils/SparseMerkleTree.sol

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.12;
+
+/**
+ * @title Sparse Merkle Tree library
+ * @notice Implements proof verification for SMTs, which is meant to be used
+ * when dealing with the checkpoint SMTs
+ * @dev This implementation uses zero as the nil node value and implements
+ * similar optimizations to those discussed in
+ * https://ethresear.ch/t/optimizing-sparse-merkle-trees/3751/4 and related
+ * articles. Proofs used here can be referred to as 'compact' proofs in some
+ * client implementations because unnecessary side nodes are omitted and
+ * indicated through a bitmask.
+ */
+library SparseMerkleTree {
+    struct Proof {
+        bytes32 leaf;
+        uint256 index;
+        uint256 bitmask;
+        bytes32[] sideNodes;
+    }
+
+    /**
+     * @notice Verifies a SMT (non-)inclusion proof, checking that the
+     * resulting root is the same as the current root
+     * @param root Current SMT root
+     * @param depth SMT depth
+     * @param proof SMT proof
+     * @return Whether the (non-)inclusion proof is valid
+     */
+    function verifyProof(bytes32 root, uint256 depth, Proof calldata proof) internal pure returns (bool) {
+        return computeRoot(depth, proof) == root;
+    }
+
+    /**
+     * @notice Computes the SMT root through an SMT (non-)inclusion proof.
+     * @param depth SMT depth
+     * @param proof SMT proof
+     * @return SMT root
+     */
+    function computeRoot(uint256 depth, Proof calldata proof) internal pure returns (bytes32) {
+        require(depth <= 256, "Depth should be at most 256");
+
+        if (depth != 256) {
+            require(proof.index < 2 ** depth, "Invalid index");
```

Comment:
Yes, but bitmask is already a `uint256` and only the relevant bits are considered. If there are more set bits than side nodes, it will lead to an index out of bounds exception when reading the side node from `sideNodes`, so should not be an issue.